packages/frontend/core/src/components/attachment-viewer/worker/worker.ts

```diff
@@ -0,0 +1,223 @@
+const ranges = {
+  [`${RenderKind.Page}:startIndex`]: 0,
+  [`${RenderKind.Page}:endIndex`]: 0,
+  [`${RenderKind.Thumbnail}:startIndex`]: 0,
+  [`${RenderKind.Thumbnail}:endIndex`]: 0,
+};
+
+function post<T extends MessageOp>(type: T, data?: MessageDataType[T]) {
+  const message = { type, [type]: data };
+  self.postMessage(message);
+}
+
+function renderToImageData(index: number, scale: number) {
+  if (!viewer || !doc) return;
+
+  const page = doc.page(index);
+
+  if (!page) return;
+
+  const width = Math.ceil(docInfo.width * scale);
+  const height = Math.ceil(docInfo.height * scale);
+
+  const bitmap = viewer.createBitmap(width, height, 0);
+  bitmap.fill(0, 0, width, height);
+  page.render(bitmap, 0, 0, width, height, 0, flags);
+
+  const data = new Uint8ClampedArray(bitmap.toBytes());
+
+  bitmap.close();
+  page.close();
+
+  return new ImageData(data, width, height);
+}
+
+function createJob(index: number, kind: RenderKind, scale: number) {
+  return () => runJob(index, kind, scale);
+}
+
+async function runJob(index: number, kind: RenderKind, scale: number) {
+  const key = `${kind}:${index}`;
+
+  let imageData = cached.size > 0 ? cached.get(key) : undefined;
+
+  if (!imageData) {
+    try {
+      imageData = renderToImageData(index, scale);
+    } catch (err) {
+      console.error(err);
+    }
+
+    if (!imageData) return;
+
+    cached.set(key, imageData);
+  }
+
+  post(MessageOp.Rendered, { index, kind, imageData });
+}
+
+function clearOut(kind: RenderKind, startIndex: number, endIndex: number) {
+  const oldStartIndex = ranges[`${kind}:startIndex`];
+  const oldEndIndex = ranges[`${kind}:endIndex`];
+  let i = 0;
+  let l = 0;
+
+  if (oldEndIndex < startIndex || oldStartIndex > endIndex) {
+    i = oldStartIndex;
+    l = oldEndIndex;
+  } else {
+    const oldMid = Math.ceil((oldStartIndex + oldEndIndex) / 2);
+    const mid = Math.ceil((startIndex + endIndex) / 2);
+    const diff = Math.abs(mid - oldMid);
+
+    if (mid > oldMid) {
+      i = oldStartIndex;
+      l = i + diff;
+    } else if (mid < oldMid) {
+      i = endIndex;
+      l = i + diff;
+    }
+  }
+
+  for (; i < l + 1; i++) {
+    cached.delete(`${kind}:${i}`);
+  }
+
+  ranges[`${kind}:startIndex`] = startIndex;
+  ranges[`${kind}:endIndex`] = endIndex;
+}
+
+async function start() {
+  inited = true;
+
+  logger.debug('pdf worker pending');
+  self.postMessage({ type: MessageOp.Init });
+
+  const pdfium = await createPDFium();
+  viewer = new Viewer(new Runtime(pdfium));
+
+  self.postMessage({ type: MessageOp.Inited });
+  logger.debug('pdf worker ready');
+}
+
+async function process({ data }: MessageEvent<MessageData>) {
+  if (!inited) {
+    await start();
+  }
+
+  if (!viewer) return;
+
+  const { type } = data;
+
+  switch (type) {
+    case MessageOp.Open: {
+      const buffer = data[type];
+      if (!buffer) return;
+
+      doc = viewer.open(new Uint8Array(buffer));
+
+      if (!doc) return;
```

Comment:
avoid make `doc` worker scope variable. I don't know what will happen if two different PDFs opened and call `Render` parallelly. it would be in a mess.

a better choice is every time `Open` message received, create a `PDFRenderer` with that PDF as context and cache it instead of rendered images. actually rendered images could be internal state of every `PDFRenderer`

Reply:
The current implementation, without opening a pdf, creates a worker instance, so there is no conflict.

Reply:
this is not good either. workers are running in standalone threads, it will run out of the resources if too many pdfs rendered at the same time